src/mesh/voxel_cloud.rs

```diff
@@ -228,6 +239,163 @@ impl VoxelCloud {
         }
     }
 
+    /// Computes boolean intersection (logical AND operation) of the current and
+    /// another Voxel cloud. The Voxel clouds have to intersect and the voxel
+    /// dimension must be equal. The current Voxel cloud will be mutated and
+    /// resized to the size and position of an intersection of the two Voxel
+    /// clouds (not the contained volumes).
+    pub fn boolean_intersection_mut(&mut self, other: &VoxelCloud) {
+        assert!(
+            approx::relative_eq!(self.voxel_dimensions, other.voxel_dimensions()),
+            "The two voxel clouds don't contain voxels of the same size"
+        );
+
+        // Find volume common to both voxel clouds.
+        // FIXME: This could be delegated to BoundingBox. @Optimization Compute
+        // bounding boxes so that they contain actual voxels.
+        let self_min = self.block_start;
+        let self_max = self.block_end();
+        let other_min = other.block_start;
+        let other_max = other.block_end();
+
+        let min_coord = Point3::new(
+            self_min.x.max(other_min.x),
+            self_min.y.max(other_min.y),
+            self_min.z.max(other_min.z),
+        );
+        let max_coord = Point3::new(
+            self_max.x.min(other_max.x),
+            self_max.y.min(other_max.y),
+            self_max.z.min(other_max.z),
+        );
+
+        // If the voxel clouds don't share the same chunk of space, there can't
+        // be any intersection.
+        assert!(
+            (max_coord.coords - min_coord.coords)
+                .iter()
+                .all(|v| *v >= 0),
+            "The two voxel clouds don't intersect"
+        );
```

Comment:
but is it? 
1. where is the `block_start` then?
2. how would it then perform `to_mesh` when we don't allow empty meshes?
3. also, if we don't allow empty meshes, why would we allow zero length voxel clouds?

Reply:
> where is the block_start then?

Tough. But since the cloud has no dimensions, we can perhaps afford to just pick something arbitrary, like the original block start. 

> how would it then perform to_mesh when we don't allow empty meshes?

`to_mesh` already returns an option for this exact reason. Well, for that and for weld too.

> also, if we don't allow empty meshes, why would we allow zero length voxel clouds?

You said earlier that two voxel clouds are equivalent if they have the same volumes, even if their bounding boxes or voxel map data is different. This same argument I think makes all the storage metadata question irrelevant. A different question would be, whether we want to allow a voxel cloud without a set volume. But that would be even more problematic.

Also. Maybe we should revisit allowing empty meshes :slightly_frowning_face: